### PR TITLE
langref: improve atomics documentation

### PR DESCRIPTION
**Pull request by wooster0**
This does a few things.
* This moves some documentation into `lib/std/builtin.zig` where it belongs.
* Resolves inconsistencies in parameter types.
  Sometimes it was `@import("std").builtin.X`, sometimes `std.builtin.X`, sometimes `builtin.X`, or just `X`.
  Now I changed it all to `std.builtin.X` because it's a good balance
  between not being too long (especially for some of the atomics builtins) while not being ambiguous
  (the `builtin` in `builtin.X` might be mistaken as `@import("builtin")` instead of `std.builtin`).
* No longer copy-pastes `PrefetchOptions` into the langref.
  I think this is just a bad idea long-term. It will get out of date
  over and over and also if we're copy-pasting `PrefetchOptions` why
  aren't we doing that for other structs? That's inconsistent.
  The user should know to reference the standard library documentation
  if they want to know what an std struct that is used in a builtin's prototype does.
  We shouldn't mix std documentation and language documentation.

### CODE REVIEW

File: lib/std/builtin.zig [hunk 1]
```diff
@@ -79,11 +79,24 @@ pub const SymbolVisibility = enum {
...
+    /// other threads performing an `Acquire` or `SeqCst` operation on the same atomic value.
     Release,
+    /// Combines the guarantees of both `Acquire` and `Release` orderings:
+    /// ensures that all memory accesses and modifications before the atomic operation are visible to the current thread,
+    /// and all memory accesses and modifications after the atomic operation are visible to other threads.
     AcqRel,
+    /// "Sequentially consistent" - provides the strongest synchronization guarantee among all memory orderings.
+    /// Ensures that all memory accesses and modifications from all threads are seen in a total order that is consistent with the program order.
+    /// This ordering establishes a global order of memory operations and ensures strong consistency across all threads.
     SeqCst,
```
Comment (r00ster91) [inline lib/std/builtin.zig:100]: This documentation is new and I'm not entirely sure about it because I don't have much experience with threading/atomics overall. CC @kprotty do you think it's accurate?

Comment (kprotty) [reply lib/std/builtin.zig:100]: Don't think so:

- Unordered means the operation is atomic but doesnt have to have an observable modification order. At least that's what I gather [from LLVM](https://llvm.org/docs/Atomics.html#unordered:~:text=It%20essentially%20guarantees%20that%20races%20produce%20somewhat%20sane%20results%20instead%20of%20having%20undefined%20behavior) in relation to C++ memory orderings.
- Monotonic enforces an [ordering of reads/writes](https://llvm.org/docs/Atomics.html#unordered:~:text=It%20essentially%20guarantees%20that%20if%20you%20take%20all%20the%20operations%20affecting%20a%20specific%20address%2C%20a%20consistent%20ordering%20exists) on its value to other threads by having a modification order (its not linked to SeqCst). All memory accesses in program order are visible to the current thread even without atomics.
- Acquire ensures that if an atomic value becomes visible through a store, and that store was Release, all memory ops before that Release store are visible to the Acquirer. It doesn't affect other threads which Release.
- SeqCst is not consistent with program order (program order isnt a global/cross-thread concept). AFAIK, it mainly provides a shared modification order across multiple atomic variables. This shared MO doesnt provide happens-before (i.e. "memory ops before become visible" like Acq/Rel) across atomic variables though.

Comment (kprotty) [reply lib/std/builtin.zig:100]: Defining the meanings for the orderings might rely on https://github.com/ziglang/zig/issues/6396 being addressed.

Comment (matu3ba) [reply lib/std/builtin.zig:100]: 1. This is missing in the langref, that these result in potential hardware fences and compiler fences (for stuff on the same thread like signal handler and interrupt handler/exceptions are not yet supported.
2. I'm not entirely sure if correct, but I think "happens before" relations and semantics in general are for now according to the C11 memory model with a note that this is planned to be changed eventually (out of thin air problem). This would also belong into the langref.
3. Might be worth to mention somewhere (langref?), that LLVM is currently not able to model and optimize low level primitives at scale and thus intrinsics should be preferred instead of inline assembly. 
Background is that there is no coherent optimization model favored by the verification community, so each backend does its own thing.

Comment (r00ster91) [reply lib/std/builtin.zig:100]: @kprotty I like your wording so I took over some of what you said into the documentation and removed some that I wasn't sure about. I think this is still a good start and before or after #6396 is addressed we can of course still adjust this further.

File: doc/langref.html.in [hunk 2]
```diff
@@ -7831,35 +7831,19 @@ comptime {
       {#header_close#}
 
       {#header_open|@atomicRmw#}
-      <pre>{#syntax#}@atomicRmw(comptime T: type, ptr: *T, comptime op: builtin.AtomicRmwOp, operand: T, comptime ordering: builtin.AtomicOrder) T{#endsyntax#}</pre>
+      <pre>{#syntax#}@atomicRmw(comptime T: type, ptr: *T, comptime op: std.builtin.AtomicRmwOp, operand: T, comptime ordering: std.builtin.AtomicOrder) T{#endsyntax#}</pre>
       <p>
-      This builtin function atomically modifies memory and then returns the previous value.
+      This builtin function atomically modifies memory and then returns the previous value - "Read-Modify-Write".
```
Comment (kprotty) [inline doc/langref.html.in:7836]: What about: "This builtin function, short for "Read Modify Write", atomically ..."

File: doc/langref.html.in [hunk 3]
```diff
@@ -7831,35 +7831,19 @@ comptime {
...
-            Also supports {#link|Floats#}.</li>
-        <li>{#syntax#}.Sub{#endsyntax#} - for integers, twos complement wraparound subtraction.
-            Also supports {#link|Floats#}.</li>
-        <li>{#syntax#}.And{#endsyntax#} - bitwise and</li>
-        <li>{#syntax#}.Nand{#endsyntax#} - bitwise nand</li>
-        <li>{#syntax#}.Or{#endsyntax#} - bitwise or</li>
-        <li>{#syntax#}.Xor{#endsyntax#} - bitwise xor</li>
-        <li>{#syntax#}.Max{#endsyntax#} - stores the operand if it is larger. Supports integers and floats.</li>
-        <li>{#syntax#}.Min{#endsyntax#} - stores the operand if it is smaller. Supports integers and floats.</li>
-      </ul>
```
Comment (kprotty) [inline doc/langref.html.in:7857]: Why was this section removed?

File: doc/langref.html.in [hunk 4]
```diff
@@ -8475,13 +8459,10 @@ export fn @"A function name that is a complete sentence."() void {}
       {#header_close#}
 
       {#header_open|@fence#}
-      <pre>{#syntax#}@fence(order: AtomicOrder) void{#endsyntax#}</pre>
+      <pre>{#syntax#}@fence(order: std.builtin.AtomicOrder) void{#endsyntax#}</pre>
       <p>
       The {#syntax#}fence{#endsyntax#} function is used to introduce happens-before edges between operations.
```
Comment (kprotty) [inline doc/langref.html.in:8464]: Maybe "between atomic operations"

File: doc/langref.html.in [hunk 5]
```diff
@@ -8899,37 +8891,6 @@ test "@wasmMemoryGrow" {
...
-        read,
-        write,
-    };
-
-    pub const Cache = enum(u1) {
-        instruction,
-        data,
-    };
-};
-      {#code_end#}
```
Comment (kprotty) [inline doc/langref.html.in:8932]: Why was this removed too?

File: lib/std/builtin.zig [hunk 6]
```diff
@@ -79,11 +79,23 @@ pub const SymbolVisibility = enum {
 /// This data structure is used by the Zig language code generation and
 /// therefore must be kept in sync with the compiler implementation.
 pub const AtomicOrder = enum {
+    /// The operation is atomic but does not have to have an observable modification order.
     Unordered,
+    /// Enforces an ordering of reads and writes on its value to other threads by having a modification order.
+    /// All memory accesses in program order are visible to the current thread.
```
Comment (kprotty) [inline lib/std/builtin.zig:85]: - "The operation is atomic but has a stable ordering of reads and writes to a given atomic variable (modification order)"
- The "program order" bit can be omitted as that's true for all Orderings, even non-atomic/normal memory operations.

**Review by kprotty — COMMENTED** [review]
Opinion: we probably shouldn't try documenting AtomicOrder until the memory model is documented. There's a few edge cases we'd have to note in addition to what's covered. If not, incorrect invariants could be inferred like "observing normal store A with Acquire, which happened after Release store B, means you get the Acq<->Rel visibility of B" (this is false AFAIK)

File: doc/langref.html.in (hunk 2 above)
Comment (wooster0) [reply doc/langref.html.in:7836]: I would rather get rid of "This builtin function" as that is completely redundant and this is one of the few places where that's done. This style of explaining an acronym or abbreviation is used for `@popCount` too:

> Counts the number of bits set in an integer - "population count".

I think in this PR I'll stick with it for now.

File: doc/langref.html.in (hunk 3 above)
Comment (wooster0) [reply doc/langref.html.in:7857]: Please see the PR description.

File: doc/langref.html.in (hunk 5 above)
Comment (wooster0) [reply doc/langref.html.in:8932]: Please see the PR description.

**Comment by wooster0** [comment]
> we probably shouldn't try documenting AtomicOrder until the memory model is documented

Fair enough. I removed the docs for now.

**Comment by rootbeer** [comment]
Not that I have any standing in Zig to bless this change, but the doc changes here look good to me.   Moving the standard library doc into the standard library itself seems like the right direction to go in, and the textual change looks legit to me.  The other cleanups to langref in here look good too.

**Comment by wooster0** [comment]
Yup, and I also just resolved the conflicts that I noticed thanks to your comment @rootbeer. So if anyone is interested in merging this...
